File: src/main/java/com/github/jaksonlin/testcraft/infrastructure/commands/unittestannotations/CheckAnnotationCommand.java
package com.github.jaksonlin.testcraft.infrastructure.commands.unittestannotations;

import com.github.jaksonlin.testcraft.domain.annotations.AnnotationFieldType;
import com.github.jaksonlin.testcraft.domain.annotations.AnnotationSchema;
import com.github.jaksonlin.testcraft.infrastructure.commands.testscan.UnittestCaseCheckCommand;
import com.github.jaksonlin.testcraft.domain.context.CaseCheckContext;
import com.github.jaksonlin.testcraft.domain.context.UnittestCase;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiAnnotation;
import com.intellij.psi.PsiComment;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.util.PsiTreeUtil;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class CheckAnnotationCommand extends UnittestCaseCheckCommand {

    public CheckAnnotationCommand(Project project, CaseCheckContext context) {
        super(project, context);
    }

    @Override
    public void execute() {
        PsiAnnotation annotation = findTargetAnnotation(getContext().getPsiMethod(), getContext().getSchema());
        if (annotation == null) {
            showNoAnnotationMessage(getProject(), getContext().getSchema().getAnnotationClassName());
            return;
        }
        processAnnotation(annotation);
    }

    private void processAnnotation(PsiAnnotation annotation) {
        try {
            UnittestCase testCase = parseUnittestCaseFromAnnotations(annotation);
            String message = formatTestCaseMessage(testCase, getContext().getSchema());
            showSuccessMessage(getProject(), message);
        } catch (Exception e) {
            showErrorMessage(getProject(), e.getMessage() != null ? e.getMessage() : "Unknown error");
        }
    }

    private String extractMethodBodyComments(PsiMethod psiMethod) {
        List<String> stepComments = new ArrayList<>();
        List<String> assertComments = new ArrayList<>();

        PsiTreeUtil.findChildrenOfType(psiMethod, PsiComment.class).forEach(comment -> {
            if (comment.getText().toLowerCase().contains("step")) {
                stepComments.add(comment.getText());
            }
            if (comment.getText().toLowerCase().contains("assert")) {
                assertComments.add(comment.getText());
            }
        });

        String stepCommentsFormatted = stepComments.stream()
                .map(comment -> {
                    int index = stepComments.indexOf(comment);
                    return "test_step_" + (index + 1) + ": " + comment.substring(2);
                })
                .collect(Collectors.joining("\n"));

        String assertCommentsFormatted = assertComments.stream()
                .map(comment -> {
                    int index = assertComments.indexOf(comment);
                    return "test_assert_" + (index + 1) + ": " + comment.substring(2);
                })
                .collect(Collectors.joining("\n"));

        return stepCommentsFormatted + (stepCommentsFormatted.isEmpty() ? "" : "\n") + assertCommentsFormatted;
    }

    private String formatTestCaseMessage(UnittestCase testCase, AnnotationSchema schema) {
        StringBuilder sb = new StringBuilder();
        sb.append("Test Case Details:\n");
        schema.getFields().forEach(field -> {
            sb.append(field.getName()).append(": ");
            if (field.getType() == AnnotationFieldType.STRING) {
                sb.append(testCase.getString(field.getName())).append("\n");
            } else if (field.getType() == AnnotationFieldType.STRING_LIST) {
                sb.append(String.join(", ", testCase.getStringList(field.getName()))).append("\n");
            }
        });
        sb.append(extractMethodBodyComments(getContext().getPsiMethod())).append("\n");
        return sb.toString();
    }
}

File: src/main/kotlin/com/github/jaksonlin/pitestintellij/commands/unittestannotations/CheckAnnotationCommand.kt
package com.github.jaksonlin.pitestintellij.commands.unittestannotations

import com.github.jaksonlin.pitestintellij.annotations.AnnotationFieldType
import com.github.jaksonlin.pitestintellij.annotations.AnnotationSchema
import com.github.jaksonlin.pitestintellij.context.CaseCheckContext
import com.github.jaksonlin.pitestintellij.context.UnittestCase
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiAnnotation
import com.intellij.psi.PsiComment
import com.intellij.psi.PsiMethod
import com.intellij.psi.util.PsiTreeUtil

class CheckAnnotationCommand(project: Project,  context: CaseCheckContext):UnittestCaseCheckCommand(project, context) {
    override fun execute() {
        val annotation = findTargetAnnotation(context.psiMethod, context.schema)
        if (annotation == null) {
            showNoAnnotationMessage(project, context.schema.annotationClassName)
            return
        }
        processAnnotation(annotation)
        return
    }



    private fun processAnnotation(
        annotation: PsiAnnotation
    ) {
        try {
            val testCase = parseUnittestCaseFromAnnotations(annotation)
            val message = formatTestCaseMessage(testCase, context.schema)
            showSuccessMessage(project, message)
        } catch (e: Exception) {
            showErrorMessage(project, e.message ?: "Unknown error")
        }
    }

    private fun extractMethodBodyComments(psiMethod: PsiMethod): String {
        val stepComments = mutableListOf<String>()
        val assertComments = mutableListOf<String>()

        PsiTreeUtil.findChildrenOfType(psiMethod, PsiComment::class.java).forEach {
            if (it.text.contains("step", ignoreCase = true)) {
                stepComments.add(it.text)
            }
            if (it.text.contains("assert", ignoreCase = true)) {
                assertComments.add(it.text)
            }
        }
        // format as : test_step_1: step_comment, remove the leading //
        return stepComments.mapIndexed { index, comment -> "test_step_${index + 1}: ${comment.substring(2)}" }.joinToString("\n") + "\n" + assertComments.mapIndexed { index, comment -> "test_assert_${index + 1}: ${comment.substring(2)}" }.joinToString("\n")
    }


    private fun formatTestCaseMessage(
        testCase: UnittestCase,
        schema: AnnotationSchema
    ): String = buildString {
        appendLine("Test Case Details:")
        schema.fields.forEach { field ->
            append(field.name)
            append(": ")
            when (field.type) {
                AnnotationFieldType.STRING ->
                    appendLine(testCase.getString(field.name))
                AnnotationFieldType.STRING_LIST ->
                    appendLine(testCase.getStringList(field.name).joinToString(", "))
            }
        }
        appendLine(extractMethodBodyComments(context.psiMethod))
    }

}